util/postprocess-makedepend.pl

```diff
@@ -0,0 +1,81 @@
+#
+# Licensed under the OpenSSL license (the "License").  You may not use
+# this file except in compliance with the License.  You can obtain a copy
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+
+my $producer = $ARGV[0];
+
+while (<STDIN>) {
+    if ($producer eq 'makedepend') {
+
+        # makedepend, in its infinite wisdom, wants to have the object file in
+        # the same directory as the source file.  This doesn't work too well
+        # with out-of-source-tree builds, so we must resort to tricks to get
+        # things right.  The trick is to call makedepend with an extra suffix
+        # that contains the desired object file path, like this:
+        #
+        #     makedepend -f- -o"|dir/foo.o" -- $(CFLAGS) -- ../somewhere/foo.c
+        #
+        # The result will look something like this:
+        #
+        #     ../somewhere/foo|dir/foo.o: deps...
+        #
+        # Which is easy to massage by removing everything up to the first |
+
+        # Remove everything up to the first |
+        s/^.*\|//;
+        # Also, remove any dependency that starts with a /, because those are
+        # typically system headers
+        s/ \/(\\.|[^ ])*//;
+        # Finally, remove all empty lines or comment lines
+        $_ = undef if (/: *$/ || /^(#.*| *)$/);
+        $_.="\n" unless !defined($_) or /\R$/g;
+
+    } elsif ($producer eq 'VMS C') {
+
+        # current versions of DEC / Compaq / HP / VSI C strips away all
+        # directory information from the object file, so we must insert it
+        # back. Just to be safe against future changes, we check that there
+        # really is no directory information.
+
+        my $directory = $ARGV[1];
+
+        # The pattern for target and dependencies will always take this form:
+        #
+        #     target SPACE : SPACE deps
+        #
+        # This is so a volume delimiter (a : without any spaces around it)
+        # won't get mixed up with the target / deps delimiter.  We use this
+        # fact in the regexp below to make sure we do look at the target.
+        s/^/$directory/ unless /^\S+[:<\]]\S+\s+:/;
+
+    } elsif ($producer eq 'Windows C') {
+
+        # For the moment, we only know of one native Windows C compiler, and
+        # that's Visual C.  With that compiler, the flags /Zs /showIncludes
+        # give us the necessary output to be able to create dependencies that
+        # nmake (or any 'make' implementation) should be able to read, with a
+        # bit of help.  The output we're interested in looks like this (it
+        # always starts the same)
+        #
+        #     Note: including file: {whatever header file}
+        #
+        # So all we really have to do is to is to replace the start of the line
+        # with an object file specification, given to us as $ARGV[1].
+
+        my $object = $ARGV[1];
+        s/^Note: including file: */${object}: /;
```

Comment:
If I look at generated file I spot things like `..\openssl\crypto\some\thing.c(100): warning C4267 ...`. In other words it looks like one should omit all non-Note lines.

Secondly it lists all the system header files, and I don't think this is actually intentional. Of course problem is how to distinguish system header files from OpenSSL ones. Trouble there is that it produces absolute paths even for files that you include with relative reference. I mean you say -I..\openssl, but /showIncludes produces absolute path. And so you have mixture of `c:\program files\windows kits\10\some\thing` and `c:\users\user\some\where\openssl`. Is it possible to use SRCDIR and BLDIR somehow?

Reply:
However! I couldn't help noticing that all those .d files ... are not used for anything, as depend: rule in the makefile is empty.

Reply:
> If I look at generated file I spot things like `..\openssl\crypto\some\thing.c(100): warning C4267 ...`. In other words it looks like one should omit all non-Note lines.

Ah, of course, that's why the previous hack had this collection thing going on.  I totally forgot!  Will fix.

> Secondly it lists all the system header files, and I don't think this is actually intentional.

I know, and I left it like that for exactly the reasons you mention, that it's difficult to distinguish them from our own.  I have the exact same kind of problem on VMS, although, the system header files should be possible to detect, 'cause they all contain a `$` and are actually text libraries (all header files enclosed), extension `.TLB`.

> Is it possible to use SRCDIR and BLDIR somehow?

It would if they were absolute, but they might not be...  they are definitely relative when on the same device.

Reply:
> However! I couldn't help noticing that all those .d files ... are not used for anything, as depend: rule in the makefile is empty.

Ah, one thing I forgot about!  Well, cmd isn't as easy to script as bash, or even DCL, so I haven't quite figured out how to do this thing on Windows.  Do you think you can lend a hand?

Reply:
> > Is it possible to use SRCDIR and BLDIR somehow?

> they are relative

And there is no Perl method to convert them to absolute?

Reply:
There is, and we can certainly construct a couple of variables that are guaranteed to hold the absolute paths.
Not by the computer now, will do tomorrow.

Reply:
As for adding dependency lists to makefile. So far perl served as substitute in such situations. There is File::Find that should be usable as `find $(DEPS) -newer Makefile`. One probably could use same script on all systems...